src/cascadia/TerminalSettingsModel/ActionMap.cpp

```diff
@@ -0,0 +1,557 @@
+            FAIL_FAST_IF_NULL(cmd);
+
+            return !cmd.HasNestedCommands() && cmd.ActionAndArgs().Action() == ShortcutAction::Invalid ?
+                       nullptr : // explicitly unbound
+                       cmd;
+        }
+
+        // We don't have an answer
+        return std::nullopt;
+    }
+
+    // Method Description:
+    // - Retrieves a map of command names to the commands themselves
+    // - These commands should not be modified directly because they may result in
+    //    an invalid state for the `ActionMap`
+    Windows::Foundation::Collections::IMapView<hstring, Model::Command> ActionMap::NameMap()
+    {
+        if (!_NameMapCache)
+        {
+            // populate _NameMapCache
+            std::unordered_map<hstring, Model::Command> nameMap{};
+            _PopulateNameMapWithNestedCommands(nameMap);
+            _PopulateNameMapWithStandardCommands(nameMap);
+
+            _NameMapCache = single_threaded_map<hstring, Model::Command>(std::move(nameMap));
+        }
+        return _NameMapCache.GetView();
+    }
+
+    // Method Description:
+    // - Populates the provided nameMap with all of our nested commands and our parents nested commands
+    // - Performs a top-down approach by going to the root first, then recursively adding the nested commands layer-by-layer
+    // Arguments:
+    // - nameMap: the nameMap we're populating. This maps the name (hstring) of a command to the command itself.
+    void ActionMap::_PopulateNameMapWithNestedCommands(std::unordered_map<hstring, Model::Command>& nameMap) const
+    {
+        // Update NameMap with our parents.
+        // Starting with this means we're doing a top-down approach.
+        FAIL_FAST_IF(_parents.size() > 1);
+        for (const auto& parent : _parents)
+        {
+            parent->_PopulateNameMapWithNestedCommands(nameMap);
+        }
+
+        // Add NestedCommands to NameMap _after_ we handle our parents.
+        // This allows us to override whatever our parents tell us.
+        for (const auto& [name, cmd] : _NestedCommands)
+        {
+            if (cmd.HasNestedCommands())
+            {
+                // add a valid cmd
+                nameMap.insert_or_assign(name, cmd);
+            }
+            else
+            {
+                // remove the invalid cmd
+                nameMap.erase(name);
+            }
+        }
+    }
+
+    // Method Description:
+    // - Populates the provided nameMap with all of our actions and our parents actions
+    //    while omitting the actions that were already added before
+    // Arguments:
+    // - nameMap: the nameMap we're populating. This maps the name (hstring) of a command to the command itself.
+    //             There should only ever by one of each command (identified by the actionID) in the nameMap.
+    void ActionMap::_PopulateNameMapWithStandardCommands(std::unordered_map<hstring, Model::Command>& nameMap) const
+    {
+        const ActionHash actionHasher{};

```

Comment:
this is kinda the thing i was talking about with the hasher just being a static function here in this file. you're not using it to implement the std::hashlike interface of any stl container.